File: sdk/javascript/examples/error-handling.ts
/**
 * Example: Error Handling
 *
 * Demonstrates how to catch and handle the different error types thrown
 * by the EKET SDK:
 *
 *   - ValidationError   — bad input / missing required fields
 *   - AuthenticationError — invalid or expired JWT
 *   - NotFoundError     — resource (task/agent) does not exist
 *   - ConflictError     — task already claimed by another agent
 *   - NetworkError      — server unreachable or no response
 *   - EketError         — base class; catch-all for SDK errors
 *
 * Run: npx ts-node sdk/javascript/examples/error-handling.ts
 */

import {
  EketClient,
  ValidationError,
  AuthenticationError,
  NotFoundError,
  ConflictError,
  NetworkError,
  EketError,
} from '../src/index.js';

const SERVER_URL = process.env.EKET_SERVER_URL ?? 'http://localhost:8080';

// ── Helper: print a structured error summary ──────────────────────────────────
function reportError(label: string, err: unknown): void {
  if (err instanceof EketError) {
    console.error(`  [${err.name}] ${label}`);
    console.error(`    code   : ${err.code}`);
    console.error(`    message: ${err.message}`);
    if (err.details) {
      console.error(`    details: ${JSON.stringify(err.details)}`);
    }
  } else if (err instanceof Error) {
    console.error(`  [Error] ${label}: ${err.message}`);
  } else {
    console.error(`  [Unknown] ${label}: ${String(err)}`);
  }
}

// ── Scenario 1: ValidationError ───────────────────────────────────────────────
async function demoValidationError(client: EketClient): Promise<void> {
  console.log('\n[1] ValidationError — missing required fields');
  try {
    // agent_type and role are required; omitting them triggers a ValidationError
    await client.registerAgent({
      // @ts-expect-error intentionally omitting required fields for demo
      specialty: 'backend',
    });
  } catch (err: unknown) {
    if (err instanceof ValidationError) {
      reportError('Registration rejected due to missing fields', err);
      console.log('  ✓ Handled ValidationError correctly');
    } else {
      reportError('Unexpected error', err);
    }
  }
}

// ── Scenario 2: AuthenticationError ──────────────────────────────────────────
async function demoAuthError(): Promise<void> {
  console.log('\n[2] AuthenticationError — invalid JWT token');
  const badClient = new EketClient({
    serverUrl: SERVER_URL,
    jwtToken: 'invalid.jwt.token',
  });

  try {
    // Any authenticated endpoint will reject the bad token
    await badClient.listAgents();
  } catch (err: unknown) {
    if (err instanceof AuthenticationError) {
      reportError('Request rejected: bad token', err);
      console.log('  ✓ Handled AuthenticationError correctly');
    } else if (err instanceof NetworkError) {
      // Server might not be running — that's fine for the demo
      reportError('Server unreachable (expected in offline demo)', err);
    } else {
      reportError('Unexpected error', err);
    }
  } finally {
    await badClient.shutdown();
  }
}

// ── Scenario 3: NotFoundError ─────────────────────────────────────────────────
async function demoNotFoundError(client: EketClient): Promise<void> {
  console.log('\n[3] NotFoundError — task does not exist');
  try {
    await client.getTask('NONEXISTENT-9999');
  } catch (err: unknown) {
    if (err instanceof NotFoundError) {
      reportError('Task not found', err);
      console.log('  ✓ Handled NotFoundError correctly');
    } else if (err instanceof NetworkError) {
      reportError('Server unreachable (expected in offline demo)', err);
    } else {
      reportError('Unexpected error', err);
    }
  }
}

// ── Scenario 4: ConflictError ─────────────────────────────────────────────────
async function demoConflictError(client: EketClient, instanceId: string): Promise<void> {
  console.log('\n[4] ConflictError — task already claimed');
  try {
    // Attempting to claim the same task twice will produce a conflict
    const tasks = await client.listTasks({ status: 'ready' });
    if (tasks.length === 0) {
      console.log('  ℹ No ready tasks — skipping conflict demo');
      return;
    }

    const taskId = tasks[0].id;
    await client.claimTask(taskId, instanceId);  // first claim
    await client.claimTask(taskId, instanceId);  // second claim → conflict
  } catch (err: unknown) {
    if (err instanceof ConflictError) {
      reportError('Task already claimed', err);
      console.log('  ✓ Handled ConflictError correctly');
    } else if (err instanceof NetworkError) {
      reportError('Server unreachable (expected in offline demo)', err);
    } else {
      reportError('Unexpected error', err);
    }
  }
}

// ── Scenario 5: NetworkError ──────────────────────────────────────────────────
async function demoNetworkError(): Promise<void> {
  console.log('\n[5] NetworkError — server not reachable');
  const offlineClient = new EketClient({
    serverUrl: 'http://localhost:19999', // nothing listening here
    timeout: 2000,
  });

  try {
    await offlineClient.healthCheck();
  } catch (err: unknown) {
    if (err instanceof NetworkError) {
      reportError('Could not reach server', err);
      console.log('  ✓ Handled NetworkError correctly');
    } else {
      reportError('Unexpected error', err);
    }
  } finally {
    await offlineClient.shutdown();
  }
}

// ── Scenario 6: Generic EketError catch-all ───────────────────────────────────
async function demoGenericEketError(client: EketClient): Promise<void> {
  console.log('\n[6] EketError catch-all — unknown task update');
  try {
    // Updating a non-existent task may throw NotFoundError (subclass of EketError)
    await client.updateTask('GHOST-0000', { status: 'done' });
  } catch (err: unknown) {
    if (err instanceof EketError) {
      reportError(`SDK error (code=${err.code})`, err);
      console.log('  ✓ Caught via EketError base class');
    } else {
      reportError('Non-SDK error', err);
    }
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  console.log('=== EKET SDK — Error Handling Examples ===');

  const client = new EketClient({ serverUrl: SERVER_URL, timeout: 5000 });

  // Register a real agent so authenticated scenarios work (if server is up)
  let instanceId: string | undefined;
  try {
    const reg = await client.registerAgent({
      agent_type: 'custom',
      role: 'slaver',
      specialty: 'qa',
    });
    instanceId = reg.instance_id;
    console.log(`\nℹ Connected to server. Instance: ${instanceId}`);
  } catch {
    console.log('\nℹ Server not available — running in offline demo mode');
  }

  await demoValidationError(client);
  await demoAuthError();
  await demoNotFoundError(client);

  if (instanceId) {
    await demoConflictError(client, instanceId);
  }

  await demoNetworkError();
  await demoGenericEketError(client);

  // Cleanup
  if (instanceId) {
    try {
      await client.deregisterAgent(instanceId);
    } catch {
      // best effort
    }
  }
  await client.shutdown();

  console.log('\n✅ Error handling demo complete!');
}

main().catch((err: unknown) => {
  const message = err instanceof Error ? err.message : String(err);
  console.error('❌ Fatal error:', message);
  process.exit(1);
});

File: sdk/python/examples/error_handling.py
#!/usr/bin/env python3
"""
Example: Error Handling

Demonstrates how to handle common EKET SDK errors gracefully.

Error types:
  - ValidationError  : invalid request data (e.g., missing required fields)
  - AuthenticationError : invalid or expired JWT token
  - NotFoundError    : resource does not exist (agent / task / PR)
  - ConflictError    : resource already claimed or duplicated
  - ServerError      : 5xx response from server
  - EketError        : base class; also covers timeout & connection failures
"""

from eket_sdk import (
    EketClient,
    AgentType,
    AgentRole,
    AgentSpecialty,
    EketError,
    AuthenticationError,
    ValidationError,
    NotFoundError,
    ConflictError,
    ServerError,
)

SERVER_URL = "http://localhost:8080"

# ---------------------------------------------------------------------------
# Helper
# ---------------------------------------------------------------------------


def section(title: str) -> None:
    print(f"\n{'=' * 60}")
    print(f"  {title}")
    print(f"{'=' * 60}")


# ---------------------------------------------------------------------------
# 1. ValidationError — server rejects bad data (or client-side guard fires)
# ---------------------------------------------------------------------------

section("1. ValidationError — missing / invalid fields")

client = EketClient(server_url=SERVER_URL)

try:
    # deregister_agent() without prior registration → client raises ValidationError
    # because instance_id is None
    client.deregister_agent()
except ValidationError as e:
    print("[CAUGHT ValidationError]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
    print(f"  details : {e.details}")

# ---------------------------------------------------------------------------
# 2. NotFoundError — resource does not exist
# ---------------------------------------------------------------------------

section("2. NotFoundError — resource not found")

try:
    # Manually set an instance_id so client doesn't raise ValidationError,
    # then attempt to fetch a non-existent agent from the server.
    client.instance_id = "non-existent-agent-id"
    client.get_agent()
except NotFoundError as e:
    print("[CAUGHT NotFoundError]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
except EketError as e:
    # Server not running → connection error falls under EketError
    print("[CAUGHT EketError — server likely offline]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
finally:
    client.instance_id = None   # reset

# ---------------------------------------------------------------------------
# 3. AuthenticationError — bad / missing JWT
# ---------------------------------------------------------------------------

section("3. AuthenticationError — invalid JWT token")

client_with_bad_token = EketClient(
    server_url=SERVER_URL,
    jwt_token="invalid.jwt.token",
)
client_with_bad_token.instance_id = "some-instance"

try:
    client_with_bad_token.send_heartbeat()
except AuthenticationError as e:
    print("[CAUGHT AuthenticationError]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
except EketError as e:
    print("[CAUGHT EketError — server likely offline]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
finally:
    client_with_bad_token.close()

# ---------------------------------------------------------------------------
# 4. ConflictError — task already claimed
# ---------------------------------------------------------------------------

section("4. ConflictError — task already claimed by another agent")

client.instance_id = "agent-abc"

try:
    client.claim_task("TASK-ALREADY-CLAIMED")
except ConflictError as e:
    print("[CAUGHT ConflictError]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
except EketError as e:
    print("[CAUGHT EketError — server likely offline]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
finally:
    client.instance_id = None

# ---------------------------------------------------------------------------
# 5. Connection / timeout — server offline
# ---------------------------------------------------------------------------

section("5. EketError — connection failure (server offline)")

offline_client = EketClient(
    server_url="http://localhost:19999",   # port that is not listening
    timeout=3,
)

try:
    offline_client.health_check()
except EketError as e:
    print("[CAUGHT EketError]")
    print(f"  code    : {e.code}")
    print(f"  message : {e.message}")
finally:
    offline_client.close()

# ---------------------------------------------------------------------------
# 6. Generic catch-all pattern (recommended in production code)
# ---------------------------------------------------------------------------

section("6. Recommended catch-all pattern")


def safe_register(server_url: str) -> None:
    """Register agent with graceful degradation."""
    c = EketClient(server_url=server_url, timeout=5)
    try:
        agent = c.register_agent(
            agent_type=AgentType.CUSTOM,
            role=AgentRole.SLAVER,
            specialty=AgentSpecialty.BACKEND,
        )
        print(f"[OK] Registered as {agent.instance_id}")

    except ValidationError as e:
        print(f"[ERROR] Bad request — fix your parameters: {e.message}")

    except AuthenticationError as e:
        print(f"[ERROR] Auth failed — check your JWT token: {e.message}")

    except ServerError as e:
        print(f"[ERROR] Server error — retry later: {e.message}")

    except EketError as e:
        # Covers: connection errors, timeouts, unexpected HTTP codes
        print(f"[ERROR] SDK error ({e.code}): {e.message}")

    finally:
        c.close()


safe_register(SERVER_URL)

# ---------------------------------------------------------------------------

print(f"\n{'=' * 60}")
print("  Error handling examples complete.")
print("  (NetworkError shown when EKET server is not running)")
print(f"{'=' * 60}\n")
